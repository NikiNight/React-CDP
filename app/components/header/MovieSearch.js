import React from 'react';
import SearchByButton from './SearchByButton';
import { connect } from 'react-redux';
import { changeFiltering, saveSearchText, requestMovies } from '../../store/actions/actions';

class MovieSearch extends React.Component {

    constructor(props) {
        super(props);
        this.toggleButtonsState = this.toggleButtonsState.bind(this);
        this.searchMovieBy = this.searchMovieBy.bind(this);
    }

    toggleButtonsState(value) {
        if(value !== this.props.searchBy) {
            this.props.changeFiltering(value);
        }
    }

    searchMovieBy() {
        const inputValue = this.input.value;
        if(inputValue.length > 0) {
            this.props.saveSearchText(inputValue);
        } else {
            this.props.saveSearchText('');
        }
        this.props.requestMovies();
    }

    render() {
        return (
            <div className="movie-search">
                <h1>Find your movie</h1>
                <input type="text" name="movie" className="movie-search__input" ref={(input) => this.input=input} />
                <div className="movie-search__buttons-line">
                    <div className="movie-search__buttons">
                        <span>Search by</span>
                        <SearchByButton isActive={this.props.searchBy=='title'} value='Title' queryValue="title" toggleButtonsState={this.toggleButtonsState} />
                        <SearchByButton isActive={this.props.searchBy=='genres'} value='Genre' queryValue="genres" toggleButtonsState={this.toggleButtonsState} />
                    </div>
                    {
                        this.props.loading
                        ? <button className="button button_search button_disabled">Loading...</button>
                        : <button className="button button_search"  onClick={this.searchMovieBy}>Search</button>
                    }
                </div>
            </div>
        )
    }
}

function mapDispatchToProps (dispatch) {
    return {
        changeFiltering: (filtering) => dispatch(changeFiltering(filtering)),
        saveSearchText: (text) => dispatch(saveSearchText(text)),
        requestMovies: (params) => dispatch(requestMovies(params))
    }
}

function mapStateToProps (state) {
  const {  query, loading } = state;
  return { search: query.search, searchBy: query.searchBy, loading: loading }
}
  
export default connect(mapStateToProps, mapDispatchToProps)(MovieSearch);