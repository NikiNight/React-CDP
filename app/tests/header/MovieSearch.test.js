import React from 'react';
import renderer from 'react-test-renderer';
import MovieSearch from './../../components/header/MovieSearch';
import {shallow, mount} from 'enzyme';

describe('Test MovieSearch component', () => {

    test('matches snapshot', () => {
        const tree = renderer.create(<MovieSearch/>)
        .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should emit toggleButtonsState', () => {
        const component = mount(<MovieSearch/>);
        component.instance().toggleButtonsState('Genre');
        expect(component.instance().state.searchBySelected).toBe('Genre');
    });

    it('should set searchValue', () => {
        const component = mount(<MovieSearch/>);
        component.find('.movie-search__input').instance().value = "test";
        component.find('.button_search').simulate('click');
        expect(component.instance().state.searchValue).toBe('test');
    });
})