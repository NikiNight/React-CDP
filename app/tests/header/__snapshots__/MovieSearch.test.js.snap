// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MovieSearch component matches snapshot 1`] = `
<div
  className="movie-search"
>
  <h1>
    Find your movie
  </h1>
  <input
    className="movie-search__input"
    name="movie"
    type="text"
  />
  <div
    className="movie-search__buttons-line"
  >
    <div
      className="movie-search__buttons"
    >
      <span>
        Search by
      </span>
      <button
        className="button button_active"
        onClick={[Function]}
        value="Title"
      >
        Title
      </button>
      <button
        className="button "
        onClick={[Function]}
        value="Genre"
      >
        Genre
      </button>
    </div>
    <button
      className="button button_search"
      onClick={[Function]}
    >
      Search
    </button>
  </div>
</div>
`;
